### PR TITLE
[Communication] Job Router Public Preview v2022-07-18

### LINKED ISSUE #18013: [Communication Services - JobRouter] REST API Review
- Service name: ACS Job Router
- Key contact for the review: Bo Gao (bga@microsoft.com)
- Whether this is a new or existing API - New API
- Whether or not your service already been deployed - yes
- A brief description of the material to be reviewed/changed - APIs (swagger model) has been reviewed with Jeffrey Richter. This PR includes  swagger definition and examples for JobRouter.

### PR DESCRIPTION
**Pull request by williamzhao87**
# Data Plane API - Pull Request
Changes for Job Router public preview

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc:
* Updated paths:

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/18013

### CODE REVIEW

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (mikekistler) [inline specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json:992]: Don't add "(Optional)" in the description as it is redundant with the `required` attribute of the parameter.

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (mikekistler) [inline specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json:1094]: Better to only support "application/json".

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (mikekistler) [inline specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json:1122]: Is the request body truly optional? What is the behavior if it is not provided?

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (mikekistler) [inline specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json:1235]: Is the request body truly optional? What is the behavior if it is not provided?

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (mikekistler) [inline specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json:1514]: Patch that supports both create and update should define both 200 and 201 response codes.

**Review by mikekistler — CHANGES_REQUESTED** [review]
I flagged a number of issues raised by the Spectral linter.

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (sarkar-rajarshi) [reply specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json]: Changed to maxpagesize

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (sarkar-rajarshi) [reply specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json]: Removed defaultValue of -1 from maxpagesize

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (sarkar-rajarshi) [reply specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json]: This has been removed

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (sarkar-rajarshi) [reply specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json]: Have removed everything besides application/json

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (sarkar-rajarshi) [reply specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json:1138]: For this particular route of unassign job, there is a single optional property that `UnassignJobRequest` wraps: 'bool? SuspendMatching'. If an empty body is provided, then internally, the optional value default to false (which is the default behaviour of JobRouter in the first place).

Similar behavior exists in other routes which accepts optional body, for e.g., /routing/jobs/{id}:cancel

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (sarkar-rajarshi) [reply specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json:1249]: Same as before. For this particular route of decline offer, there is an optional property that `DeclineJobOfferRequest` wraps: `DateTimeOffset? RetryOfferAt`. If an empty body is provided, then internally, it would mean that the worker that declined the offer will not get any further offers for that particular job. If however, Contoso sets a value of `retryOfferAt`, then the worker will again get an offer for the same job after the given timestamp.

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (sarkar-rajarshi) [reply specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json]: This has been addressed. All date-time properties have "At" suffixes and "Utc" suffixes have been dropped

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (sarkar-rajarshi) [reply specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json:1533]: This has been corrected.

**Review by mikekistler — APPROVED** [review]
Thanks for the changes! Looks good. 👍

**Comment by sarkar-rajarshi** [comment]
@vicancy Could you please me help merge this?

File: specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json
Comment (vicancy) [inline specification/communication/data-plane/JobRouter/preview/2022-07-18-preview/communicationservicejobrouter.json:1672]: Rename to `state` to be consistent with the query name?
```suggestion
            "description": "If specified, select workers by worker state.",
```

**Review by vicancy — APPROVED** [review]

**Comment by sarkar-rajarshi** [comment]
Hi @tjprescott , can you help merge this?

**Comment by tjprescott** [comment]
@sarkar-rajarshi I can't merge this as long as the Swagger BreakingChange and Swagger ModelValidation checks are failing.

**Comment by sarkar-rajarshi** [comment]
@tjprescott I am working with the ARB team to get an approval on the breaking change. Meanwhile, could you help me with why the model validation is failing? I know  [READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/Semantic-and-Model-Violations-Reference.md#READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST) is failing because of the "kind" discriminator value being passed on in the request. However, I am not sure what would be way out of it? It is indeed expected in the request, but  the value of it is fixed. Its not clear what the best representation of that is in swagger.

And for [INVALID_TYPE](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/Semantic-and-Model-Violations-Reference.md#INVALID_TYPE) this is caused because of empty class definitions. The empty classes are kept in place so that introducing new props would be non-breaking. Once again, its not clear what would be way out of it?

**Comment by tjprescott** [comment]
> @tjprescott I am working with the ARB team to get an approval on the breaking change. Meanwhile, could you help me with why the model validation is failing? I know [READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/Semantic-and-Model-Violations-Reference.md#READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST) is failing because of the "kind" discriminator value being passed on in the request. However, I am not sure what would be way out of it? It is indeed expected in the request, but the value of it is fixed. Its not clear what the best representation of that is in swagger.

You should be able to just remove the kind property from the request payloads.

**Comment by tjprescott** [comment]
For the TYPE_ERROR, this may be a false positive or there may be a better way to express these empty classes. Please reach out in the following channel for guidance:
https://aka.ms/azsdk/support/specreview-channel

**Comment by sarkar-rajarshi** [comment]
> > @tjprescott I am working with the ARB team to get an approval on the breaking change. Meanwhile, could you help me with why the model validation is failing? I know [READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/Semantic-and-Model-Violations-Reference.md#READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST) is failing because of the "kind" discriminator value being passed on in the request. However, I am not sure what would be way out of it? It is indeed expected in the request, but the value of it is fixed. Its not clear what the best representation of that is in swagger.
> 
> You should be able to just remove the kind property from the request payloads.

@tjprescott I tried removing the `kind` property from the request. It now throws a [OBJECT_MISSING_REQUIRED_PROPERTY](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/Semantic-and-Model-Violations-Reference.md#OBJECT_MISSING_REQUIRED_PROPERTY) stating `Missing required property: kind` and a subsequence [DISCRIMINATOR_VALUE_NOT_FOUND](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/Semantic-and-Model-Violations-Reference.md#DISCRIMINATOR_VALUE_NOT_FOUND) -- both of which makes sense since even though the `kind` property is read-only, its a constant value that will need to passed in the request.

https://github.com/Azure/azure-rest-api-specs/pull/23804/checks?check_run_id=14833989143

**Comment by tjprescott** [comment]
> @tjprescott I tried removing the kind property from the request. It now throws a [OBJECT_MISSING_REQUIRED_PROPERTY](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/Semantic-and-Model-Violations-Reference.md#OBJECT_MISSING_REQUIRED_PROPERTY) stating Missing required property: kind and a subsequence [DISCRIMINATOR_VALUE_NOT_FOUND](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/Semantic-and-Model-Violations-Reference.md#DISCRIMINATOR_VALUE_NOT_FOUND) -- both of which makes sense since even though the kind property is read-only, its a constant value that will need to passed in the request.

cc @johanste

**Comment by johanste** [comment]
Constant string values are modeled as single-valued enums. However, I suspect that the intent is the set of valid `kind` values will grow over time, in which case I'd make this just a string (I believe there may be other heuristics a single-value extensible enum may run afoul of).

**Comment by sarkar-rajarshi** [comment]
> Constant string values are modeled as single-valued enums. However, I suspect that the intent is the set of valid `kind` values will grow over time, in which case I'd make this just a string (I believe there may be other heuristics a single-value extensible enum may run afoul of).

@tjprescott as per the feedbacl, I made `kind` just a string. As expected the [READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/Semantic-and-Model-Violations-Reference.md#READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST) validation error has gone.

I am currently still blocked on [INVALID_TYPE](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/Semantic-and-Model-Violations-Reference.md#INVALID_TYPE) for empty classes. If this is not a blocker and is a false positive, could this be overridden?

**Review by scbedd — APPROVED** [review]
Approving to unblock failing Model Validation per @lmazuel recommendation.

**Comment by sarkar-rajarshi** [comment]
@tjprescott ready to be merged

**Pull request merged**
